src/Icicle/Internal/EditDistance.hs

```diff
@@ -0,0 +1,65 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+module Icicle.Internal.EditDistance (
+    editDistance
+    ) where
+
+import           P
+import qualified Data.Vector.Unboxed.Mutable as Vector
+import           Icicle.Internal.Pretty
+
+import           Control.Monad.ST (runST, ST)
+import qualified Data.Map as Map
+import           Data.List ((!!))
+
+editDistance :: Doc -> Doc -> Int
+editDistance a b =
+  let pretty_comp = (flip displayS "") . renderCompact . pretty
+      a' = pretty_comp a
+      b' = pretty_comp b
+  in damerauLevenshtein a' b'
+
+-- This function it from the library language-spelling, but it has dependencies on
+-- ListLike, listlike instances and a version of Data.Text which clashes with what
+-- we're using. For just one function, it's not too terrible to just bring it into
+-- this file.
+
+damerauLevenshtein :: (Ord a) => [a] -> [a] -> Int
```

Comment:
FTR Thomas Sutton has presented a nice way of implementing the [edit distance algorithm at FPSyd](https://speakerdeck.com/thsutton/dynamic-programming-in-haskell.) a few months ago. The generic implementation for dynamic programming problems fits on one slide (slide 13) and uses `Vector` to explore solutions (the code is available [here](https://github.com/thsutton/talk-dynamic-programming/tree/master/haskell) if you want to have a look at it).